### PR TITLE
24 lähtötietoihin valmiiksi hallinnolliset rajat

### LINKED ISSUE #24: Lähtötietoihin valmiiksi hallinnolliset rajat.
QGIS työtilaan voisi tuoda valmiiksi MML:n rajapinnat joissa hallinnolliset rajat eri mittakaavatasoilla. 

Hallinnolliset rajat joita ainakin tarvitaan: kunnat, maakuntarajat.

Tietokantaan tarvitaan hallinnolliset rajat, jotta voidaan toteuttaa validointisääntö.

Kysytty Sykeltä mikä mittakaavan kuntaraja on käytössä, ei vielä vastattu 14.5.2024

### PR DESCRIPTION
**Pull request by msorvoja**
Fix #24

Still missing tests.

@Rikuoja One test fails since the revision file has been edited by hand. This is because alembic overwrites an optional field containing geometry to `nullable=False`. That won't do since the code list containing the names of the administrative regions must be loaded into the administrative_region code table before the geometries can be inserted. Hence, the geometry field is momentarily filled with NULL values.

### CODE REVIEW

File: database/test/test_services.py [hunk 1]
```diff
@@ -78,7 +92,9 @@ def test_create_db(create_db, main_db_params_with_root_user):
         conn.close()
 
 
-def test_populate_koodistot(populate_koodistot, main_db_params):
+def test_populate_koodistot(
+    populate_koodistot, populate_admin_region_geometries, main_db_params
+):
```
Comment (Rikuoja) [inline database/test/test_services.py:97]: Here we could perhaps make a separate test test_populate_admin_region_geometries that also uses populate_koodistot and checks that all the geometries are filled.

In that case, we would not need populate_admin_region_geometries here.

**Review by Rikuoja — CHANGES_REQUESTED** [review]
Looks good, just some minor changes!

File: database/mml_loader/mml_loader.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        output_dir = "admin_region_geom_data"
+        os.makedirs(output_dir, exist_ok=True)
+
+        year = str(self.payload["inputs"]["yearInput"])  # type: ignore
+        size = str(self.payload["inputs"]["dataSetInput"]).split("_")[-1]  # type: ignore # noqa
+
+        url = f"{self.api_base}/execution?api-key={self.api_key}"
+        LOGGER.info(f"Starting OGC API process on {self.api_base}/execution")
+        r = session.post(url, headers=self.HEADERS, json=self.payload)
+        r.raise_for_status()
```
[the remark this answers is not in the document]
Comment (Rikuoja) [reply database/mml_loader/mml_loader.py:87]: Might also be caused by rate limiting. In this case, just try this out later with a sleep of some seconds and check if it still fails all the time?

**Comment by msorvoja** [comment]
@Rikuoja Declaring geom field as `geom = Column(Geometry(geometry_type="MULTIPOLYGON", srid=3067), nullable=True)` instead of `geom: Mapped[Optional[MultiPolygon]]` fixed the problem where alembic forced the field to be non-nullable. A bug in alembic perhaps?

File: database/test/test_services.py [hunk 3]
```diff
@@ -167,6 +181,24 @@ def test_populate_local_koodistot(populate_local_koodistot, main_db_params):
...
+    Test that maakunta geometries are populated
+    """
+    conn = psycopg2.connect(**main_db_params)
+    try:
+        with conn.cursor() as cur:
+            cur.execute(
+                "SELECT count(*) FROM codes.administrative_region WHERE geom IS NULL"
+            )
+            geom_count = cur.fetchone()[0]
+            assert geom_count == 19
```
Comment (Rikuoja) [inline database/test/test_services.py:197]: @msorvoja Here I'm wondering about this test, it doesn't seem to work. Here we test that geom is NULL in all the 19 rows, it should be NOT NULL I think?

Also, when running the test, my call responds with
```
{'statusCode': 200, 'body': '"0 inserted or updated. 0 deleted."'}
```
so it looks here we are testing that the code does NOT fill any geometries.

**Review by Rikuoja — CHANGES_REQUESTED** [review]
One test question.

File: database/test/test_services.py (hunk 3 above)
Comment (msorvoja) [reply database/test/test_services.py:197]: Good catch!

**Review by Rikuoja — APPROVED** [review]

**Pull request merged**